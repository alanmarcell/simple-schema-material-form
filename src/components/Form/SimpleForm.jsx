import React from 'react';
import PropTypes from 'prop-types';
import { omit, pathOr, merge } from 'ramda';
import { compose, withStateHandlers, pure } from 'recompact';
import { Button, TextField } from 'material-ui';

const getChildProps = ({ doc, setDoc }) => Child => {
  const { fieldName } = Child.props;
  const value = pathOr('', [fieldName], doc);

  const childProps = {
    ...Child.props,
    setDoc,
    doc,
    onChange: e => setDoc({ [fieldName]: e.target.value }),
    value,
    label: fieldName,
  };

  return React.createElement(
    Child.type,
    childProps,
    null,
  );
};

export const TextInput = props => {
  const { doc, setDoc, fieldName } = props;
  const value = pathOr('', [fieldName], doc);
  const cProps = omit(['fieldName', 'setDoc', 'doc'], props);

  const inputProps = {
    ...cProps,
    onChange: e => setDoc({ [fieldName]: e.target.value }),
    value,
    label: fieldName,
  };
  return <TextField {...inputProps} />;
};

/** Our Simple Form Test Utility */
const SimpleForm = (props) => {
  const { doc, setDoc, onSubmit } = props;
  const setChildProps = getChildProps({ doc, setDoc });
  return (
    <form >
      {React.Children.map(props.children, setChildProps)}
      <Button onClick={() => onSubmit(doc)} >Submit</Button>
    </form>
  );
};

/** This is our base enhancer to handle doc changes */
const enhance = compose(
  pure,
  withStateHandlers(
    /** this our initial doc */
    { doc: {} },
    {
      /** this is the doc state handler */
      setDoc: ({ doc }) => value => {
        const merged = merge(doc, value);
        return {
          doc: merged,
        };
      },
    },
  ),
);

const EnhancedSimpleForm = enhance(SimpleForm);

SimpleForm.defaultProps = {
  /** This is our fields */
  // children: null,
};

SimpleForm.propTypes = {
  /** doc defines the form state */
  doc: PropTypes.object.isRequired,

  /** handle the form state */
  setDoc: PropTypes.func.isRequired,

  children: PropTypes.array.isRequired,
};

export default EnhancedSimpleForm;
