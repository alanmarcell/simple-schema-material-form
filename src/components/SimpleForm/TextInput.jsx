import React from 'react';
// import PropTypes from 'prop-types';
import { omit, pathOr } from 'ramda';
// import { compose, withStateHandlers, pure } from 'recompact';
import { TextField } from 'material-ui';


const TextInput = props => {
  const { doc, setDoc, fieldName } = props;
  const value = pathOr('', [fieldName], doc);
  const cProps = omit(['fieldName', 'setDoc', 'doc'], props);

  const inputProps = {
    ...cProps,
    onChange: e => setDoc({ [fieldName]: e.target.value }),
    value,
    label: fieldName,
  };
  return <TextField {...inputProps} />;
};

export default TextInput;
