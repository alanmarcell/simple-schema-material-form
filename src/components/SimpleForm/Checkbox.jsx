import React from 'react';
import { merge } from 'ramda';
import MaterialCheckbox from 'material-ui/Checkbox';
import { FormControl, FormLabel, FormControlLabel, FormGroup } from 'material-ui/Form';

const Checkbox = props => {
  const {
    setDoc, label, selectOptions = [], fieldName, doc,
  } = props;

  return (
    <FormControl >
      <FormLabel>{label}</FormLabel>
      <FormGroup row>
        {selectOptions.map(i =>
          (<FormControlLabel
            key={i.key}
            control={<MaterialCheckbox
              checked={i.checked}
              onChange={(_event, val) => setDoc(merge(doc, { [fieldName]: { ...doc[fieldName], [i.key]: val } }))}
              value={i.key || ''}
              id={i.key}
            />}
            label={i.label}
          />))
        }
      </FormGroup>
    </FormControl >
  );
};

// const Checkbox = props => {
//   const inputProps = {
//     ...props,
//   };
//   return <CheckboxField selectOptions={[{ label: 1, key: 'um' }, { label: 2, key: 'dois' }]} {...inputProps} />;
// };

export default Checkbox;
