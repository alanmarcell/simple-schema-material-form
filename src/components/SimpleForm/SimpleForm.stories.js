import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import SimpleForm from './SimpleForm';
import TextInput from '../TextInput';
import Checkbox from './Checkbox';
import withTests from '../../stories/withTests';

storiesOf('SimpleForm', module)
  .addDecorator(withTests('SimpleForm'))
  .add('with TextInput', () => (
    <SimpleForm onSubmit={action('doc')} >
      <TextInput fieldName="test" />
    </SimpleForm>
  ))
  .add('with 2 TextInput', () => (
    <SimpleForm onSubmit={action('doc')} >
      <TextInput fieldName="test" />
      <TextInput fieldName="test2" />
    </SimpleForm>
  ))
  .add('with Checkbox', () => (
    <SimpleForm onSubmit={action('doc')} >
      <Checkbox selectOptions={[{ label: 1, key: 'um' }, { label: 2, key: 'dois' }]} fieldName="test" />
    </SimpleForm>
  ))
  .add('with Checkbox and Text Iput', () => (
    <SimpleForm onSubmit={action('doc')} >
      <Checkbox selectOptions={[{ label: 1, key: 'um' }, { label: 2, key: 'dois' }]} fieldName="test" />
      <TextInput fieldName="test2" />
    </SimpleForm>
  ));

