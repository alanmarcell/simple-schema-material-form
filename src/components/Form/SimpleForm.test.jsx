import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { Button } from 'material-ui';
import SimpleForm, { TextInput } from './SimpleForm';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<SimpleForm ><TextInput fieldName="test" /></SimpleForm>, div);
  ReactDOM.unmountComponentAtNode(div);
});


it('shallow a TextInput', () => {
  const onSubmit = jest.fn();
  const SimpleFormComponent = (
    <SimpleForm onSubmit={doc => onSubmit(doc)}>
      <TextInput fieldName="test" />
    </SimpleForm>);

  const component = shallow(SimpleFormComponent);

  const newValue = 'Type Sample';
  const initialInput = component.find({ label: 'test' });
  // const initialFormState = component.state();
  // const { test } = initialFormState.childProps.doc;

  // expect(test).toBe('');

  initialInput.simulate('change', {
    target: { value: newValue },
  });

  const newState = component.state();
  const { doc } = newState.childProps;

  const newInput = component.find({ label: 'test' });
  const test2 = newState.childProps.doc.test;

  expect(test2).toBe(newValue);
  expect(newInput.props().value).toBe(newValue);
  expect(component).toMatchSnapshot();


  const button = component.find(Button);
  button.simulate('click');

  expect(onSubmit).toHaveBeenCalled();
  expect(onSubmit).toBeCalledWith(doc);
});


it('shallow a 2 TextInput', () => {
  const onSubmit = jest.fn();
  const SimpleFormComponent = (
    <SimpleForm onSubmit={doc => onSubmit(doc)}>
      <TextInput fieldName="test" />
      <TextInput fieldName="test2" />
    </SimpleForm>);

  const component = shallow(SimpleFormComponent);

  const newValue = 'Type Sample';
  const initialInput = component.find({ label: 'test' });
  const initialInput2 = component.find({ label: 'test2' });
  // const initialFormState = component.state();
  // const { test } = initialFormState.childProps.doc;

  // expect(test).toBe('');

  initialInput.simulate('change', {
    target: { value: newValue },
  });
  initialInput2.simulate('change', {
    target: { value: newValue },
  });
  const newState = component.state();
  const { doc } = newState.childProps;

  const newInput = component.find({ label: 'test' });
  const newInput2 = component.find({ label: 'test2' });

  const button = component.find(Button);
  button.simulate('click');

  const test2 = newState.childProps.doc.test;

  const expectedDoc = {
    test: newValue,
    test2: newValue,
  };

  expect(doc).toEqual(expectedDoc);

  expect(test2).toBe(newValue);
  expect(newInput.props().value).toBe(newValue);
  expect(newInput2.props().value).toBe(newValue);
  expect(component).toMatchSnapshot();
  expect(onSubmit).toHaveBeenCalled();
  expect(onSubmit).toBeCalledWith(doc);
});

