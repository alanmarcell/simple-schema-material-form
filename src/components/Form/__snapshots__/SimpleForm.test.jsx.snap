// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallow a 2 TextInput 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <pure(withStateHandlers(SimpleForm))
    onSubmit={[Function]}
>
    <Unknown
        fieldName="test"
    />
    <Unknown
        fieldName="test2"
    />
</pure(withStateHandlers(SimpleForm))>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          <Unknown
            doc={
                        Object {
                                    "test": "Type Sample",
                                    "test2": "Type Sample",
                                  }
            }
            fieldName="test"
            label="test"
            onChange={[Function]}
            setDoc={[Function]}
            value="Type Sample"
/>,
          <Unknown
            doc={
                        Object {
                                    "test": "Type Sample",
                                    "test2": "Type Sample",
                                  }
            }
            fieldName="test2"
            label="test2"
            onChange={[Function]}
            setDoc={[Function]}
            value="Type Sample"
/>,
        ],
        <WithStyles(Button)
          onClick={[Function]}
>
          Submit
</WithStyles(Button)>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": ".0",
        "nodeType": "function",
        "props": Object {
          "children": null,
          "doc": Object {
            "test": "Type Sample",
            "test2": "Type Sample",
          },
          "fieldName": "test",
          "label": "test",
          "onChange": [Function],
          "setDoc": [Function],
          "value": "Type Sample",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": ".1",
        "nodeType": "function",
        "props": Object {
          "children": null,
          "doc": Object {
            "test": "Type Sample",
            "test2": "Type Sample",
          },
          "fieldName": "test2",
          "label": "test2",
          "onChange": [Function],
          "setDoc": [Function],
          "value": "Type Sample",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Submit",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Submit",
        "type": [Function],
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            <Unknown
              doc={
                            Object {
                                          "test": "Type Sample",
                                          "test2": "Type Sample",
                                        }
              }
              fieldName="test"
              label="test"
              onChange={[Function]}
              setDoc={[Function]}
              value="Type Sample"
/>,
            <Unknown
              doc={
                            Object {
                                          "test": "Type Sample",
                                          "test2": "Type Sample",
                                        }
              }
              fieldName="test2"
              label="test2"
              onChange={[Function]}
              setDoc={[Function]}
              value="Type Sample"
/>,
          ],
          <WithStyles(Button)
            onClick={[Function]}
>
            Submit
</WithStyles(Button)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": ".0",
          "nodeType": "function",
          "props": Object {
            "children": null,
            "doc": Object {
              "test": "Type Sample",
              "test2": "Type Sample",
            },
            "fieldName": "test",
            "label": "test",
            "onChange": [Function],
            "setDoc": [Function],
            "value": "Type Sample",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": ".1",
          "nodeType": "function",
          "props": Object {
            "children": null,
            "doc": Object {
              "test": "Type Sample",
              "test2": "Type Sample",
            },
            "fieldName": "test2",
            "label": "test2",
            "onChange": [Function],
            "setDoc": [Function],
            "value": "Type Sample",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Submit",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Submit",
          "type": [Function],
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`shallow a TextInput 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <pure(withStateHandlers(SimpleForm))
    onSubmit={[Function]}
>
    <Unknown
        fieldName="test"
    />
</pure(withStateHandlers(SimpleForm))>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          <Unknown
            doc={
                        Object {
                                    "test": "Type Sample",
                                  }
            }
            fieldName="test"
            label="test"
            onChange={[Function]}
            setDoc={[Function]}
            value="Type Sample"
/>,
        ],
        <WithStyles(Button)
          onClick={[Function]}
>
          Submit
</WithStyles(Button)>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": ".0",
        "nodeType": "function",
        "props": Object {
          "children": null,
          "doc": Object {
            "test": "Type Sample",
          },
          "fieldName": "test",
          "label": "test",
          "onChange": [Function],
          "setDoc": [Function],
          "value": "Type Sample",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Submit",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Submit",
        "type": [Function],
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            <Unknown
              doc={
                            Object {
                                          "test": "Type Sample",
                                        }
              }
              fieldName="test"
              label="test"
              onChange={[Function]}
              setDoc={[Function]}
              value="Type Sample"
/>,
          ],
          <WithStyles(Button)
            onClick={[Function]}
>
            Submit
</WithStyles(Button)>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": ".0",
          "nodeType": "function",
          "props": Object {
            "children": null,
            "doc": Object {
              "test": "Type Sample",
            },
            "fieldName": "test",
            "label": "test",
            "onChange": [Function],
            "setDoc": [Function],
            "value": "Type Sample",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Submit",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Submit",
          "type": [Function],
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
